### ISSUE TITLE
Some questions about computing the results after training

### ISSUE THREAD

**Issue by yixiaoshenghua** [BensonRen/BDIMNNA#2]
Dear authors, I have run your codes to reproduce the results represented in your paper, but I cannot get the same results as yours. I just download the codes and did not change any parameters. I found the loss is a little different from yours. Is anything needed to modify when training or after training? I just run the evaluate.py after training the models. And can you tell me how to calculate the loss of the model on the meta-material dataset because it does not create a file of Ypred? The figure of the loss histogram is empty. Thanks a lot!

**Comment by BensonRen** [comment]
Hi, thank you for your interest in our work.

1. Would you mind specifying the results (dataset-algorithm pair) that you can not reproduce? One thing to note is that some of the algorithms (especially INN) are pretty unstable during training and hence need more training instances to get a good training result. That is why we put 10 as the number of training that we want and choose from those 10 trained models according to the evaluation results.

2. The meta-material evaluation does need one more step. As the documentation suggests, you need to go to the NA/predict.py folder and run the appropriate command (if you are doing single evaluation where T=1, uncomment the second chunk of code in the __main__ function, if doing multi_eval uncomment the third chunk and remember to comment the "creat_mm_dataset()"

**Comment by yixiaoshenghua** [comment]
Thank you for your answer! But I still have some questions. 
1. I have retrained each model on each dataset 10 times and can find 10 pairs of Ypred and Ytruth files in /data/ folder after evaluating. When I get the results in the fold /data/, should I compute the average value of 10 or just choose the best performance? Now, I just compute the mse between Ypred and Ytruth for each pair on each dataset and get the average value of 10 pairs. Is it right?
2. I have uncommented the second chunk of code of the main function in NA/predict.py and run it. But it is still empty in Ypred.csv for meta-material and the mse of it is nan. And I find line 280 in class_wrapper.py write that if the dataset is meta-material just pass it. Does it cause the problem I mentioned? How can I get the loss of the meta-material? 
Thank you!

**Comment by yixiaoshenghua** [comment]
Thank you for your answer! But I still have some questions. 
1. I have retrained each model on each dataset 10 times and can find 10 pairs of Ypred and Ytruth files in /data/ folder after evaluating. When I get the results in the fold /data/, should I compute the average value of 10 or just choose the best performance? Now, I just compute the mse between Ypred and Ytruth for each pair on each dataset and get the average value of 10 pairs. Is it right?
2. I have uncommented the second chunk of code of the main function in NA/predict.py and run it. But it is still empty in Ypred.csv for meta-material and the mse of it is nan. And I find line 280 in class_wrapper.py write that if the dataset is meta-material just pass it. Does it cause the problem I mentioned? How can I get the loss of the meta-material? 
Thank you!

**Comment by BensonRen** [comment]
1. Training 10 models and choose the best model from those is what I meant. Sorry for the confusion in the documentation, I would update a flow chart to provide more clarity on those.
2. I had a small update two days ago as I found in predict.py there is a mistake of "/data/tmp" where it should be only "data/", pls check if it is that problem. Would you mind emailing me if there are more things pop up, we can potentially meet online for further problems. sr365atdukedotcom

**Comment by yixiaoshenghua** [comment]
Thank you very much for your answer! I will try it again.

**Comment by JungsonStone** [comment]
> 2\. I have uncommented the second chunk of code of the main function in NA/predict.py and run it. But it is still empty in Ypred.csv for meta-material and the mse of it is nan. And I find line 280 in class_wrapper.py write that if the dataset is meta-material just pass it. Does it cause the problem I mentioned? How can I get the loss of the meta-material

I also meet this question.  This is the missing of Ypred.csv. I spend so much time that can't address this issue.  who can recover the result of authors,  please tell me.  I felt so bad. it is so difficult for the freshman to learning.

**Comment by BensonRen** [comment]
> > 2. I have uncommented the second chunk of code of the main function in NA/predict.py and run it. But it is still empty in Ypred.csv for meta-material and the mse of it is nan. And I find line 280 in class_wrapper.py write that if the dataset is meta-material just pass it. Does it cause the problem I mentioned? How can I get the loss of the meta-material
> 
> I also meet this question. This is the missing of Ypred.csv. I spend so much time that can't address this issue. who can recover the result of authors, please tell me. I felt so bad. it is so difficult for the freshman to learning.

Sorry for causing the frustration, would you mind providing more details of which step were you stuck and couldn't get Ypred.csv? Which method-dataset pair were you using etc.?

**Comment by JungsonStone** [comment]
thanks for your reply.  first, I just want to recover this figure 3 in the paper(arXiv:2009.12919v3 ). unfortunately, to run the NA/evaluate.py find this error that it is empty in test_Ypred_retrain0meta_material.csv for meta-material when I follow the flow chart process.  then, I read the note that 'if you have the meta-material dataset, do the extra step of predict.py to get Ypred ', but I don't process this dataset, what should I do to make the test_Ypred_retrain0meta_material.csv keeping the number? it is confused for me.  in a word, I don't know how to produce the file Ypredxxx.csv that making the procedure go on.  Looking forward to your answer. Thank you.

**Comment by BensonRen** [comment]
Hi thanks for the explanation of the issue you encountered. 

There are some part in your question that I still did not quite understand like "but I don't process this dataset, what should I do to make the test_Ypred_retrain0meta_material.csv keeping the number?"

Let me rephrase your question to check my understanding:
You followed the flow chart, reached the stage called "Evaluate (multi)" and you did evaluated those methods (There are files output in the multi_eval/ folder). However, when you are plotting them, there is an error occuring that "test_Ypred_retrain0meta_material.csv" file is empty?

Couple of things I would double check:
1. Did you run the `utils/delete.py` first before you do the `NA/predict.py` for meta-material dataset (Strictly following the flow chart)
2. Did you run the` NA/predict.py` line 206-209 (starting from : # Multi evaluation in the multi_eval folder of each method) to get the correct Ypred?

After you run the utils/delete.py, you should only have the below files in your `BDIMNNA/multi_eval/NA/meta_material 
1. test_Xpred_pointretrain0meta_materialinference{0-200).csv
2. Xpred.csv
3. Xtruth.csv
4. Ytruth.csv

and there should be no files called "test_Ypred_retrain0meta_material.csv" after you run the `utils/delete.py`.

Please let me know if you have any other questions and I am very happy to help you on resovling any of those.

**Comment by JungsonStone** [comment]
thank you！
